src/ffi.lisp

```diff
@@ -913,122 +912,99 @@ Note: the _really_ old formats (<= 0.9.4) are not supported."
+(defun io-wait (stream fd deadline direction)
+  "Signals `ssl-timeout' if the underlying stream timeout is hit or ssl-stream-read-timeout
+or ssl-stream-write-timeout is specified."
+  (declare (ignore fd))
+  (unless
+      (system:wait-for-input-streams (list (ssl-stream-socket stream))
+                                     :timeout (or (get-timeout stream direction)
+                                                  (get-timeout-from-deadline deadline))
+                                     :wait-reason (if (eq direction :input)
+                                                      "cl+ssl waiting for input"
+                                                      "cl+ssl waiting for output"))
+    (signal-if-timeout-expired stream direction)))
 
 #+clozure-common-lisp
-(defun input-wait (stream fd deadline)
-  (unless deadline
-    (setf deadline (stream-deadline (ssl-stream-socket stream))))
-  (let* ((timeout
-          (if deadline
-              (milliseconds-until-deadline deadline stream)
-              nil)))
-    (multiple-value-bind (win timedout error)
-        (ccl::process-input-wait fd timeout)
-      (unless win
-        (if timedout
-            (error 'ccl::communication-deadline-expired :stream stream)
-            (ccl::stream-io-error stream (- error) "read"))))))
-
-#+sbcl
-(defun input-wait (stream fd deadline)
-  (declare (ignore stream))
-  (let ((timeout
-         ;; *deadline* is handled by wait-until-fd-usable automatically,
-         ;; but we need to turn a user-specified deadline into a timeout
-         (when deadline
-           (seconds-until-deadline deadline))))
-    (sb-sys:wait-until-fd-usable fd :input timeout)))
-
-#+allegro
-(defun input-wait (stream fd deadline)
-  (declare (ignore stream))
-  (let ((timeout
-         (when deadline
-           (max 0 (seconds-until-deadline deadline)))))
-    (mp:wait-for-input-available fd
-                                 :timeout timeout
-                                 :whostate "cl+ssl waiting for input")))
-
-#+lispworks
-(defun input-wait (stream fd deadline)
-  (declare (ignore fd))
+(defun milliseconds-until-deadline (deadline stream)
+  (let* ((now (get-internal-real-time)))
+    (if (> now deadline)
+        (error 'ccl::communication-deadline-expired :stream stream)
+        (values
+         (round (- deadline now) (/ internal-time-units-per-second 1000))))))
 
-  (let* ((timeout
-           (when deadline
-             (max 0 (seconds-until-deadline deadline)))))
-    (system:wait-for-input-streams (list (ssl-stream-socket stream))
-                                   :timeout timeout
-                                   :wait-reason "cl+ssl waiting for input")))
+#+clozure-common-lisp
+(defun io-wait (stream fd deadline direction)
+  "Supports old behavior that throws `ccl::communication-deadline-expired. if the underlying stream
+   is the source of the deadline.  New behavior invoked by ssl-stream-read-timeout or write-timeout
+   throws `ssl-timeout'"
```

Comment:
I agree that the old behaviour should only be invoked by old interface, and new interface should produce the new behavior. But the `milliseconds-until-deadline` definition above signals the old, CCL-specific condition in all situations, right?